File: lib/core/services/bookmark_service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:clashkingapp/core/services/api_service.dart';
import 'package:clashkingapp/features/clan/models/clan.dart';
import 'package:clashkingapp/features/player/models/player.dart';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;

class BookmarkService extends ChangeNotifier {
  BookmarkService({ApiService? apiService})
    : _apiService = apiService ?? ApiService() {
    unawaited(load());
  }

  final ApiService _apiService;
  bool _loaded = false;
  int _loadGeneration = 0;
  String? _currentUserId;
  List<BookmarkedPlayer> _players = [];
  List<BookmarkedClan> _clans = [];

  bool get loaded => _loaded;
  List<BookmarkedPlayer> get players => List.unmodifiable(_players);
  List<BookmarkedClan> get clans => List.unmodifiable(_clans);

  void setCurrentUserId(String? userId) {
    final normalizedUserId = userId?.trim();
    final nextUserId = normalizedUserId == null || normalizedUserId.isEmpty
        ? null
        : normalizedUserId;
    if (_currentUserId == nextUserId) return;
    _currentUserId = nextUserId;
    _invalidateLoadedState();
  }

  void _invalidateLoadedState() {
    _loadGeneration++;
    _loaded = false;
  }

  Future<void> load() async {
    final loadGeneration = ++_loadGeneration;
    if (!_hasCurrentUser) {
      if (loadGeneration != _loadGeneration) return;
      _players = [];
      _clans = [];
      _loaded = false;
      return;
    }

    final userId = Uri.encodeComponent(_currentUserId!);
    final responses = await Future.wait([
      _apiService.getResponse(
        '/links/$userId/bookmarks?type=player',
        requiresAuth: true,
      ),
      _apiService.getResponse(
        '/links/$userId/bookmarks?type=clan',
        requiresAuth: true,
      ),
    ]);

    final players = _decodeBookmarkItems(
      responses[0],
    ).map(BookmarkedPlayer.fromApiJson).toList();
    final clans = _decodeBookmarkItems(
      responses[1],
    ).map(BookmarkedClan.fromApiJson).toList();
    if (loadGeneration != _loadGeneration) return;
    _players = players;
    _clans = clans;
    _loaded = true;
    notifyListeners();
  }

  bool get _hasCurrentUser {
    final userId = _currentUserId;
    return userId != null && userId.isNotEmpty;
  }

  List<Map<String, dynamic>> _decodeBookmarkItems(http.Response response) {
    // The remote bookmarks endpoint is optional/newer API surface. A 404 means
    // the user has no server-side bookmark resource yet, or the deployed API
    // does not expose this typed route. Bookmarks should not block app startup.
    if (response.statusCode == 404) {
      return const [];
    }

    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw HttpException(
        'Failed to load bookmarks (${response.statusCode})',
        uri: response.request?.url,
      );
    }

    final data = jsonDecode(ApiService.decodeResponseBody(response));
    if (data is! Map<String, dynamic> || data['items'] is! List) {
      throw const FormatException('Invalid bookmarks payload');
    }

    return (data['items'] as List).whereType<Map<String, dynamic>>().toList();
  }

  bool isPlayerBookmarked(String tag) {
    return _players.any((player) => player.tag == tag);
  }

  bool isClanBookmarked(String tag) {
    return _clans.any((clan) => clan.tag == tag);
  }

  Future<void> togglePlayer(Player player) async {
    if (isPlayerBookmarked(player.tag)) {
      await removePlayer(player.tag);
    } else {
      await addPlayer(BookmarkedPlayer.fromPlayer(player));
    }
  }

  Future<void> addPlayer(BookmarkedPlayer player) async {
    final previous = List<BookmarkedPlayer>.from(_players);
    _players.removeWhere((existing) => existing.tag == player.tag);
    _players.insert(0, player);
    notifyListeners();

    try {
      if (_hasCurrentUser) {
        await _createBookmark('player', player.tag);
      } else {
        throw UnauthorizedException('User not authenticated');
      }
    } catch (_) {
      _players = previous;
      notifyListeners();
      rethrow;
    }
  }

  Future<void> removePlayer(String tag) async {
    final previous = List<BookmarkedPlayer>.from(_players);
    _players.removeWhere((player) => player.tag == tag);
    notifyListeners();

    try {
      if (_hasCurrentUser) {
        await _deleteBookmark('player', tag);
      } else {
        throw UnauthorizedException('User not authenticated');
      }
    } catch (_) {
      _players = previous;
      notifyListeners();
      rethrow;
    }
  }

  Future<void> reorderPlayer(int oldIndex, int newIndex) async {
    if (oldIndex < 0 || oldIndex >= _players.length) return;
    if (newIndex < 0 || newIndex > _players.length) return;
    final previous = List<BookmarkedPlayer>.from(_players);
    final player = _players.removeAt(oldIndex);
    _players.insert(newIndex, player);
    notifyListeners();

    try {
      if (_hasCurrentUser) {
        await _saveBookmarkOrder(
          'player',
          _players.map((player) => player.tag).toList(growable: false),
        );
      } else {
        throw UnauthorizedException('User not authenticated');
      }
    } catch (_) {
      _players = previous;
      notifyListeners();
      rethrow;
    }
  }

  Future<void> toggleClan(Clan clan) async {
    if (isClanBookmarked(clan.tag)) {
      await removeClan(clan.tag);
    } else {
      await addClan(BookmarkedClan.fromClan(clan));
    }
  }

  Future<void> addClan(BookmarkedClan clan) async {
    final previous = List<BookmarkedClan>.from(_clans);
    _clans.removeWhere((existing) => existing.tag == clan.tag);
    _clans.insert(0, clan);
    notifyListeners();

    try {
      if (_hasCurrentUser) {
        await _createBookmark('clan', clan.tag);
      } else {
        throw UnauthorizedException('User not authenticated');
      }
    } catch (_) {
      _clans = previous;
      notifyListeners();
      rethrow;
    }
  }

  Future<void> removeClan(String tag) async {
    final previous = List<BookmarkedClan>.from(_clans);
    _clans.removeWhere((clan) => clan.tag == tag);
    notifyListeners();

    try {
      if (_hasCurrentUser) {
        await _deleteBookmark('clan', tag);
      } else {
        throw UnauthorizedException('User not authenticated');
      }
    } catch (_) {
      _clans = previous;
      notifyListeners();
      rethrow;
    }
  }

  Future<void> reorderClan(int oldIndex, int newIndex) async {
    if (oldIndex < 0 || oldIndex >= _clans.length) return;
    if (newIndex < 0 || newIndex > _clans.length) return;
    final previous = List<BookmarkedClan>.from(_clans);
    final clan = _clans.removeAt(oldIndex);
    _clans.insert(newIndex, clan);
    notifyListeners();

    try {
      if (_hasCurrentUser) {
        await _saveBookmarkOrder(
          'clan',
          _clans.map((clan) => clan.tag).toList(growable: false),
        );
      } else {
        throw UnauthorizedException('User not authenticated');
      }
    } catch (_) {
      _clans = previous;
      notifyListeners();
      rethrow;
    }
  }

  Future<void> _createBookmark(String type, String tag) async {
    final response = await _apiService.postResponse(
      _bookmarksEndpoint(),
      body: {'type': type, 'tag': tag},
      requiresAuth: true,
    );
    _throwOnBookmarkFailure(response, 'create');
  }

  Future<void> _deleteBookmark(String type, String tag) async {
    final response = await _apiService.deleteResponse(
      _bookmarkItemEndpoint(type, tag),
      requiresAuth: true,
    );
    _throwOnBookmarkFailure(response, 'delete');
  }

  Future<void> _saveBookmarkOrder(String type, List<String> tags) async {
    final response = await _apiService.putResponse(
      '${_bookmarksEndpoint()}/order',
      body: {'type': type, 'ordered_tags': tags},
      requiresAuth: true,
    );
    _throwOnBookmarkFailure(response, 'reorder');
  }

  String _bookmarksEndpoint() {
    final userId = Uri.encodeComponent(_currentUserId ?? '');
    return '/links/$userId/bookmarks';
  }

  String _bookmarkItemEndpoint(String type, String tag) {
    return '${_bookmarksEndpoint()}/$type/${Uri.encodeComponent(tag)}';
  }

  void _throwOnBookmarkFailure(http.Response response, String action) {
    if (response.statusCode >= 200 && response.statusCode < 300) return;

    throw HttpException(
      'Failed to $action bookmark (${response.statusCode})',
      uri: response.request?.url,
    );
  }
}

class BookmarkedPlayer {
  const BookmarkedPlayer({
    required this.tag,
    required this.name,
    required this.townHallLevel,
    required this.townHallPic,
    required this.clanTag,
    required this.clanName,
    required this.trophies,
    required this.league,
    required this.leagueUrl,
  });

  final String tag;
  final String name;
  final int townHallLevel;
  final String townHallPic;
  final String clanTag;
  final String clanName;
  final int trophies;
  final String league;
  final String leagueUrl;

  factory BookmarkedPlayer.fromPlayer(Player player) {
    return BookmarkedPlayer(
      tag: player.tag,
      name: player.name,
      townHallLevel: player.townHallLevel,
      townHallPic: player.townHallPic,
      clanTag: player.clanTag,
      clanName: player.clan?.name ?? player.clanOverview.name,
      trophies: player.trophies,
      league: player.league,
      leagueUrl: player.leagueUrl,
    );
  }

  factory BookmarkedPlayer.fromApiJson(Map<String, dynamic> json) {
    final tag = json['player_tag']?.toString() ?? json['tag']?.toString() ?? '';
    return BookmarkedPlayer(
      tag: tag,
      name: tag.isEmpty ? 'Unknown Player' : tag,
      townHallLevel: 0,
      townHallPic: '',
      clanTag: '',
      clanName: '',
      trophies: 0,
      league: '',
      leagueUrl: '',
    );
  }
}

class BookmarkedClan {
  const BookmarkedClan({
    required this.tag,
    required this.name,
    required this.badgeUrl,
    required this.clanLevel,
    required this.memberCount,
  });

  final String tag;
  final String name;
  final String badgeUrl;
  final int clanLevel;
  final int memberCount;

  factory BookmarkedClan.fromClan(Clan clan) {
    return BookmarkedClan(
      tag: clan.tag,
      name: clan.name,
      badgeUrl: clan.badgeUrls.medium,
      clanLevel: clan.clanLevel,
      memberCount: clan.members,
    );
  }

  factory BookmarkedClan.fromApiJson(Map<String, dynamic> json) {
    final tag = json['clan_tag']?.toString() ?? json['tag']?.toString() ?? '';
    return BookmarkedClan(
      tag: tag,
      name: tag.isEmpty ? 'Unknown Clan' : tag,
      badgeUrl: '',
      clanLevel: 0,
      memberCount: 0,
    );
  }
}

File: expo/src/core/bookmarks/bookmark-service.ts
import {
  BookmarksAddEndpoint,
  BookmarksDeleteEndpoint,
  BookmarksListEndpoint,
  BookmarksOrderEndpoint,
} from '@clashking/api-contracts/expo';
import { ApiResponseError, ResponseDecodeError } from '@clashking/api-client';
import { Effect } from 'effect';

import { UnauthorizedException, type ContractApiService } from '../api/contract-api';
import type { Clan } from '../../features/clan/models';
import type { Player } from '../../features/player/models/player';

export type BookmarkType = 'player' | 'clan';
export type BookmarkListener = () => void;

export class BookmarkHttpException extends Error {
  constructor(
    readonly status: number,
    readonly action: 'load' | 'create' | 'delete' | 'reorder',
    readonly url: string,
  ) {
    super(
      action === 'load'
        ? `Failed to load bookmarks (${status})`
        : `Failed to ${action} bookmark (${status})`,
    );
    this.name = 'BookmarkHttpException';
  }
}

export class BookmarkFormatException extends Error {
  constructor() {
    super('Invalid bookmarks payload');
    this.name = 'BookmarkFormatException';
  }
}

export class BookmarkedPlayer {
  constructor(
    readonly tag: string,
    readonly name: string,
    readonly townHallLevel: number,
    readonly townHallPic: string,
    readonly clanTag: string,
    readonly clanName: string,
    readonly trophies: number,
    readonly league: string,
    readonly leagueUrl: string,
  ) {}

  static fromPlayer(player: Player): BookmarkedPlayer {
    const linkedClan = isRecord(player.clan) ? player.clan : null;
    return new BookmarkedPlayer(
      player.tag,
      player.name,
      player.townHallLevel,
      player.townHallPic,
      player.clanTag,
      typeof linkedClan?.name === 'string' ? linkedClan.name : player.clanOverview.name,
      player.trophies,
      player.league,
      player.leagueUrl,
    );
  }

  static fromApiJson(json: Record<string, unknown>): BookmarkedPlayer {
    const tag = String(json.player_tag ?? json.tag ?? '');
    return new BookmarkedPlayer(tag, tag.length ? tag : 'Unknown Player', 0, '', '', '', 0, '', '');
  }
}

export class BookmarkedClan {
  constructor(
    readonly tag: string,
    readonly name: string,
    readonly badgeUrl: string,
    readonly clanLevel: number,
    readonly memberCount: number,
  ) {}

  static fromClan(clan: Clan): BookmarkedClan {
    return new BookmarkedClan(
      clan.tag,
      clan.name,
      clan.badgeUrls.smallest,
      clan.clanLevel,
      clan.members,
    );
  }

  static fromApiJson(json: Record<string, unknown>): BookmarkedClan {
    const tag = String(json.clan_tag ?? json.tag ?? '');
    return new BookmarkedClan(tag, tag.length ? tag : 'Unknown Clan', '', 0, 0);
  }
}

export class BookmarkService {
  private hasLoaded = false;
  private loadGeneration = 0;
  private currentUserId: string | null = null;
  private playerBookmarks: BookmarkedPlayer[] = [];
  private clanBookmarks: BookmarkedClan[] = [];
  private readonly listeners = new Set<BookmarkListener>();
  private disposed = false;

  constructor(private readonly api: ContractApiService) {}

  get loaded(): boolean {
    return this.hasLoaded;
  }

  get players(): readonly BookmarkedPlayer[] {
    return Object.freeze([...this.playerBookmarks]);
  }

  get clans(): readonly BookmarkedClan[] {
    return Object.freeze([...this.clanBookmarks]);
  }

  subscribe(listener: BookmarkListener): () => void {
    if (this.disposed) return () => undefined;
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  dispose(): void {
    this.disposed = true;
    this.listeners.clear();
  }

  setCurrentUserId(userId: string | null | undefined): void {
    const normalized = userId?.trim() ?? '';
    const nextUserId = normalized.length ? normalized : null;
    if (this.currentUserId === nextUserId) return;
    this.currentUserId = nextUserId;
    this.loadGeneration += 1;
    this.hasLoaded = false;
  }

  async load(): Promise<void> {
    const generation = ++this.loadGeneration;
    if (!this.hasCurrentUser) {
      if (generation !== this.loadGeneration) return;
      this.playerBookmarks = [];
      this.clanBookmarks = [];
      this.hasLoaded = false;
      return;
    }

    const [playerResponse, clanResponse] = await Promise.all([
      this.listBookmarks('player'),
      this.listBookmarks('clan'),
    ]);
    const players = this.decodeBookmarkItems(playerResponse).map(BookmarkedPlayer.fromApiJson);
    const clans = this.decodeBookmarkItems(clanResponse).map(BookmarkedClan.fromApiJson);
    if (generation !== this.loadGeneration) return;
    this.playerBookmarks = players;
    this.clanBookmarks = clans;
    this.hasLoaded = true;
    this.notify();
  }

  isPlayerBookmarked(tag: string): boolean {
    return this.playerBookmarks.some((player) => player.tag === tag);
  }

  isClanBookmarked(tag: string): boolean {
    return this.clanBookmarks.some((clan) => clan.tag === tag);
  }

  async togglePlayer(player: Player): Promise<void> {
    if (this.isPlayerBookmarked(player.tag)) await this.removePlayer(player.tag);
    else await this.addPlayer(BookmarkedPlayer.fromPlayer(player));
  }

  async addPlayer(player: BookmarkedPlayer): Promise<void> {
    const previous = [...this.playerBookmarks];
    this.playerBookmarks = [
      player,
      ...this.playerBookmarks.filter((item) => item.tag !== player.tag),
    ];
    this.notify();
    try {
      this.requireCurrentUser();
      await this.createBookmark('player', player.tag);
    } catch (error) {
      this.playerBookmarks = previous;
      this.notify();
      throw error;
    }
  }

  async removePlayer(tag: string): Promise<void> {
    const previous = [...this.playerBookmarks];
    this.playerBookmarks = this.playerBookmarks.filter((player) => player.tag !== tag);
    this.notify();
    try {
      this.requireCurrentUser();
      await this.deleteBookmark('player', tag);
    } catch (error) {
      this.playerBookmarks = previous;
      this.notify();
      throw error;
    }
  }

  async reorderPlayer(oldIndex: number, newIndex: number): Promise<void> {
    if (oldIndex < 0 || oldIndex >= this.playerBookmarks.length) return;
    if (newIndex < 0 || newIndex > this.playerBookmarks.length) return;
    const previous = [...this.playerBookmarks];
    const reordered = [...this.playerBookmarks];
    const [player] = reordered.splice(oldIndex, 1);
    this.playerBookmarks = reordered;
    if (newIndex > reordered.length) {
      // Flutter removes first and then List.insert throws outside its rollback
      // block for an original-length destination index.
      throw new RangeError('newIndex is outside the post-removal list.');
    }
    reordered.splice(newIndex, 0, player!);
    this.notify();
    try {
      this.requireCurrentUser();
      await this.saveBookmarkOrder(
        'player',
        reordered.map((item) => item.tag),
      );
    } catch (error) {
      this.playerBookmarks = previous;
      this.notify();
      throw error;
    }
  }

  async reorderPlayers(orderedTags: readonly string[]): Promise<void> {
    const normalized = orderedTags.map((tag) => tag.trim().toUpperCase());
    if (new Set(normalized).size !== normalized.length) {
      throw new RangeError('Player bookmark order contains duplicate tags.');
    }
    const requested = new Map(
      this.playerBookmarks.map((player) => [player.tag.trim().toUpperCase(), player]),
    );
    const visibleOrder = normalized.flatMap((tag) => {
      const player = requested.get(tag);
      return player ? [player] : [];
    });
    const visibleTags = new Set(visibleOrder.map((player) => player.tag.trim().toUpperCase()));
    let visibleIndex = 0;
    const reordered = this.playerBookmarks.map((player) =>
      visibleTags.has(player.tag.trim().toUpperCase()) ? visibleOrder[visibleIndex++]! : player,
    );
    if (reordered.every((player, index) => player === this.playerBookmarks[index])) return;

    const previous = [...this.playerBookmarks];
    this.playerBookmarks = reordered;
    this.notify();
    try {
      this.requireCurrentUser();
      await this.saveBookmarkOrder(
        'player',
        reordered.map((player) => player.tag),
      );
    } catch (error) {
      this.playerBookmarks = previous;
      this.notify();
      throw error;
    }
  }

  async toggleClan(clan: Clan): Promise<void> {
    if (this.isClanBookmarked(clan.tag)) await this.removeClan(clan.tag);
    else await this.addClan(BookmarkedClan.fromClan(clan));
  }

  async addClan(clan: BookmarkedClan): Promise<void> {
    const previous = [...this.clanBookmarks];
    this.clanBookmarks = [clan, ...this.clanBookmarks.filter((item) => item.tag !== clan.tag)];
    this.notify();
    try {
      this.requireCurrentUser();
      await this.createBookmark('clan', clan.tag);
    } catch (error) {
      this.clanBookmarks = previous;
      this.notify();
      throw error;
    }
  }

  async removeClan(tag: string): Promise<void> {
    const previous = [...this.clanBookmarks];
    this.clanBookmarks = this.clanBookmarks.filter((clan) => clan.tag !== tag);
    this.notify();
    try {
      this.requireCurrentUser();
      await this.deleteBookmark('clan', tag);
    } catch (error) {
      this.clanBookmarks = previous;
      this.notify();
      throw error;
    }
  }

  async reorderClan(oldIndex: number, newIndex: number): Promise<void> {
    if (oldIndex < 0 || oldIndex >= this.clanBookmarks.length) return;
    if (newIndex < 0 || newIndex > this.clanBookmarks.length) return;
    const previous = [...this.clanBookmarks];
    const reordered = [...this.clanBookmarks];
    const [clan] = reordered.splice(oldIndex, 1);
    this.clanBookmarks = reordered;
    if (newIndex > reordered.length) {
      throw new RangeError('newIndex is outside the post-removal list.');
    }
    reordered.splice(newIndex, 0, clan!);
    this.notify();
    try {
      this.requireCurrentUser();
      await this.saveBookmarkOrder(
        'clan',
        reordered.map((item) => item.tag),
      );
    } catch (error) {
      this.clanBookmarks = previous;
      this.notify();
      throw error;
    }
  }

  private get hasCurrentUser(): boolean {
    return this.currentUserId !== null && this.currentUserId.length > 0;
  }

  private requireCurrentUser(): string {
    if (!this.hasCurrentUser) throw new UnauthorizedException('User not authenticated');
    return this.currentUserId!;
  }

  private decodeBookmarkItems(response: Awaited<ReturnType<typeof this.listBookmarks>>): readonly Record<string, unknown>[] {
    if (!response.ok) return [];
    return response.value.items;
  }

  private async createBookmark(type: BookmarkType, tag: string): Promise<void> {
    const userId = this.requireCurrentUser();
    await this.bookmarkRequest(
      Effect.runPromise(this.api.execute(BookmarksAddEndpoint, { path: { userId }, query: {}, body: { type, tag } })),
      'create',
      BookmarksAddEndpoint.path,
    );
  }

  private async deleteBookmark(type: BookmarkType, tag: string): Promise<void> {
    const userId = this.requireCurrentUser();
    const response = await this.bookmarkRequest(
      Effect.runPromise(this.api.executeStatus(BookmarksDeleteEndpoint, { path: { userId, type, tag }, query: {}, body: {} })),
      'delete',
      BookmarksDeleteEndpoint.path,
    );
    if (!response.ok) throw new BookmarkHttpException(response.status, 'delete', BookmarksDeleteEndpoint.path);
  }

  private async saveBookmarkOrder(type: BookmarkType, tags: readonly string[]): Promise<void> {
    const userId = this.requireCurrentUser();
    await this.bookmarkRequest(
      Effect.runPromise(this.api.execute(BookmarksOrderEndpoint, { path: { userId }, query: {}, body: { type, ordered_tags: [...tags] } })),
      'reorder',
      BookmarksOrderEndpoint.path,
    );
  }

  private listBookmarks(type: BookmarkType) {
    const userId = this.requireCurrentUser();
    return this.bookmarkRequest(
      Effect.runPromise(this.api.executeStatus(BookmarksListEndpoint, { path: { userId }, query: { type }, body: {} })),
      'load',
      BookmarksListEndpoint.path,
    );
  }

  private async bookmarkRequest<T>(
    request: Promise<T>,
    action: 'load' | 'create' | 'delete' | 'reorder',
    path: string,
  ): Promise<T> {
    try {
      return await request;
    } catch (error) {
      if (error instanceof ResponseDecodeError) throw new BookmarkFormatException();
      if (error instanceof ApiResponseError) {
        throw new BookmarkHttpException(error.status, action, path);
      }
      throw error;
    }
  }

  private notify(): void {
    if (!this.disposed) for (const listener of this.listeners) listener();
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}
